certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -435,19 +435,19 @@ def _rank_matches_by_name_and_ssl(self, vhost_list, target_name):
             if name_type == 'exact':
                 matches.append({'vhost': vhost,
                                 'name': name,
-                                'rank': 0 if vhost.ssl else 1})
+                                'rank': 0 if vhost.ssl else 4})
             elif name_type == 'wildcard_start':
                 matches.append({'vhost': vhost,
                                 'name': name,
-                                'rank': 2 if vhost.ssl else 3})
+                                'rank': 1 if vhost.ssl else 5})
             elif name_type == 'wildcard_end':
                 matches.append({'vhost': vhost,
                                 'name': name,
-                                'rank': 4 if vhost.ssl else 5})
+                                'rank': 2 if vhost.ssl else 6})
             elif name_type == 'regex':
                 matches.append({'vhost': vhost,
                                 'name': name,
-                                'rank': 6 if vhost.ssl else 7})
+                                'rank': 3 if vhost.ssl else 7})
         return sorted(matches, key=lambda x: x['rank'])
```

Comment:
Previous reasoning (imagined, since I wasn't there for it): since we're going to enable ssl anyway, might as well pick the most name-matching one on any port. then after, we'll be ssl *and* as name-matching as possible

Updated reasoning: follow the user's stated preferences that already exist in the config, matching nginx's server block selection algorithm as closely as possible, then only falling back to having to make a vhost ssl if we absolutely have to because there isn't already one. except not quite absolutely, because it'll still choose the `default` block or first block listening on that port if there's no `default` block for that port, but still a better happy medium given that we intend people to have a name-matching non-ssl block before they start out.

Since this is the only place we rank, this is the only place where the preference is expressed, and all other vhost-choosers will be updated to use the same ranking.

Reply:
`_select_best_name_match` does need to be fixed to match the ranks though